### DISCUSSION TITLE
How to connect directly to the database?

### DISCUSSION THREAD

**Discussion by Gizmo1337** [appwrite/appwrite#4957] [Q&A] [answered]
### 🔖 Feature description

 **An easy, streamlined, and accessible connection string displayed in the console for the underlying MariaDB (or whatever other DB would be in the future) would help a million, for my needs**, and maybe anyone with my use case. I did struggle for a few hours to access it by going into the container from docker and extrapolating the correct connection info that would work with a network bridge.

### 🎤 Pitch

It's really hard to use the out-of-the-box NoSQL wrapper to do more complex operations, like data aggregations or reporting. Not sure how AppWrite organizes data.

It was even more obvious as I am also running a supabase app, and the feature of having this connection info streamlined is really helpful, as I am running a separate BE that has some cron jobs that just query the DB and post-process reports.

![image](https://user-images.githubusercontent.com/25569867/210538423-481b0e0e-ba3e-4bfb-a474-6f199515dc55.png)

### 👀 Have you spent some time to check if this issue has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/appwrite/appwrite/blob/HEAD/CODE_OF_CONDUCT.md)

**Comment by stnguyen90** [comment, marked as answer]
@Gizmo1337 thanks for raising this question! 🙏🏼 

First off, we highly discourage connecting directly to the database for core functionality of your app because:

1. you lose out on caching or the cache might become out of sync if you write directly to the database
2. you lose out on the permissions system
3. Appwrite is designed as an abstraction layer on top of the underlying database

That said, if you want to connect to the underlying database, you can do so just like any other docker container. In short, you can:

1. forward the port and then connect using the IP Address/hostname of your server and port
    * be sure to use a secure password and lock down network access before opening up the port
    * to forward the port, I recommend creating a `docker-compose.override.yml` file if you don't have one and adding a forward for mariadb like:
       * ```yml
          services:
            mariadb:
              ports:
                - "3306:3306"
          ```
3. exec into the container and connect to the db
    * I use [this](https://gist.github.com/stnguyen90/ffe40b7fe85985fc80f6bc9e0d430b3d) script I use to connect to my appwrite mariadb database

**Reply by Gizmo1337** [reply]
Thank you for the time saver! I do enjoy AppWrite I just needed for data-reading purposes and making decisions about some data aggregations, and reports. All of those are just too cumbersome with a collection-document structure.

**Comment by eldadfux** [comment]
We also have this container on the development version that gives you a browser UI to access the database when self-hosting: https://github.com/appwrite/appwrite/blob/master/docker-compose.yml#L766-L774
